File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
import re

def solution(str1, str2):
    str1 = [str1[i:i+2].lower() for i in range(len(str1)-1) if re.match('[a-zA-Z]{2}', str1[i:i+2])]
    str2 = [str2[i:i+2].lower() for i in range(len(str2)-1) if re.match('[a-zA-Z]{2}', str2[i:i+2])]

    intersection = set(str1) & set(str2)
    union = set(str1) | set(str2)

    if len(union) == 0: return 65536

    intersection = sum(min(str1.count(v), str2.count(v)) for v in intersection)
    union = sum(max(str1.count(v), str2.count(v)) for v in union)

    return int((intersection/union) * 65536)

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.js
function solution (str1, str2) {

    function slice_str(s) {
        const arr = [];
        for (let i = 0; i < s.length - 1; i++) {
            const str = s.slice(i, i+2);
            if (str.match(/[A-Za-z]{2}/)) arr.push(str.toLowerCase());
        }
        return arr;
    }

    const arr1 = slice_str(str1);
    const arr2 = slice_str(str2);
    
    const set = new Set([...arr1, ...arr2]);
    
    let union = 0;
    let intersection = 0;

    for (let v of set){
        const c1 = arr1.filter(x => x === v).length;
        const c2 = arr2.filter(x => x === v).length;
        union += Math.max(c1, c2);
        intersection += Math.min(c1, c2);
    }
    
    return union === 0 ? 65536 : Math.floor(intersection / union * 65536);
}